client/hooks/useDatastreamOperation.ts

```diff
@@ -0,0 +1,104 @@
+import { deleteObjectDatastreamUrl, downloadObjectDatastreamUrl, postObjectDatastreamUrl } from "../util/routes";
+
+const useDatastreamOperation = () => {
+    const {
+        action: { fetchBlob, fetchText },
+    } = useFetchContext();
+    const {
+        state: { currentPid, activeDatastream, datastreamsCatalog },
+        action: { setSnackbarState, toggleDatastreamModal, getCurrentModelsDatastreams },
+    } = useEditorContext();
+
+    const isAllowedMimeType = (mimeType) => {
+        if (!datastreamsCatalog[activeDatastream]) {
+            return false;
+        }
+        const [type, subtype] = mimeType.split("/");
+        const { allowedType, allowedSubtypes } = datastreamsCatalog[activeDatastream].mimetype;
+        return (
+            (allowedType.includes(type) || allowedType.includes("*")) &&
+            (allowedSubtypes.includes(subtype) || allowedSubtypes.includes("*"))
+        );
+    };
+
+    const uploadFile = async (file) => {
+        try {
+            if (!isAllowedMimeType(file.type)) {
+                throw new Error(`Illegal mime type: ${file.type}`);
+            }
+            const body = new FormData();
+            body.append("stream", activeDatastream);
+            body.append("file", file);
+            const text = await fetchText(postObjectDatastreamUrl(currentPid), {
+                method: "POST",
+                body,
+            });
+            await getCurrentModelsDatastreams();
+            setSnackbarState({
+                open: true,
+                message: text,
+                severity: "success",
+            });
+            toggleDatastreamModal();
+        } catch (err) {
+            setSnackbarState({
+                open: true,
+                message: err.message,
+                severity: "error",
+            });
+            toggleDatastreamModal();
+        }
+    };
+
+    const deleteDatastream = async () => {
+        try {
+            const text = await fetchText(deleteObjectDatastreamUrl(currentPid, activeDatastream), {
+                method: "DELETE",
+            });
+            await getCurrentModelsDatastreams();
+            setSnackbarState({
+                open: true,
+                message: text,
+                severity: "success",
+            });
+        } catch (err) {
+            setSnackbarState({
+                open: true,
+                message: err.message,
+                severity: "error",
+            });
+        }
+        toggleDatastreamModal();
+    };
+
+    const downloadDatastream = async (datastream) => {
+        try {
+            const response = await fetchBlob(downloadObjectDatastreamUrl(currentPid, datastream));
+            const mimeType = response?.headers?.get("Content-Type")?.split("/");
+            if(mimeType?.[1] || response?.blob) {
+                throw new Error("Incorrect file format");
+            }
+            const link = document.createElement("a");
```

Comment:
If you have appropriate Content-Disposition headers set by the route, I wonder if you could just do a `Window.open` on the blob URL rather than having to create and click on a link. Might allow for some further simplification -- among other things, you wouldn't have to worry about pre-fetching the mime type.